src/solvers/dg/dg_2d_indicators.jl

```diff
@@ -0,0 +1,219 @@
+      right = cache.interfaces.neighbor_ids[2, interface]
+
+      # Apply smoothing
+      alpha[left]  = max(alpha_tmp[left],  0.5 * alpha_tmp[right], alpha[left])
+      alpha[right] = max(alpha_tmp[right], 0.5 * alpha_tmp[left],  alpha[right])
+    end
+
+    # Loop over L2 mortars
+    for mortar in eachmortar(dg, cache)
+      # Get neighboring element ids
+      lower = cache.mortars.neighbor_ids[1, mortar]
+      upper = cache.mortars.neighbor_ids[2, mortar]
+      large = cache.mortars.neighbor_ids[3, mortar]
+
+      # Apply smoothing
+      alpha[lower] = max(alpha_tmp[lower], 0.5 * alpha_tmp[large], alpha[lower])
+      alpha[upper] = max(alpha_tmp[upper], 0.5 * alpha_tmp[large], alpha[upper])
+      alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[lower], alpha[large])
+      alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[upper], alpha[large])
+    end
+  end
+
+  return alpha
+end
+
+
+
+# this method is used when the indicator is constructed as for shock-capturing volume integrals
+function create_cache(::Type{IndicatorLöhner}, equations::AbstractEquations{2}, basis::LobattoLegendreBasis)
+
+  alpha = Vector{real(basis)}()
+
+  A = Array{real(basis), ndims(equations)}
+  indicator_threaded = [A(undef, nnodes(basis), nnodes(basis)) for _ in 1:Threads.nthreads()]
+
+  return (; alpha, indicator_threaded)
+end
+
+# this method is used when the indicator is constructed as for AMR
+function create_cache(typ::Type{IndicatorLöhner}, mesh, equations::AbstractEquations{2}, dg::DGSEM, cache)
+  create_cache(typ, equations, dg.basis)
+end
+
+
+function (löhner::IndicatorLöhner)(u::AbstractArray{<:Any,4}, equations, dg::DGSEM, cache)
+  @assert nnodes(dg) >= 3 "IndicatorLöhner only works for nnodes >= 3 (polydeg > 1)"
+  @unpack alpha, indicator_threaded = löhner.cache
+  resize!(alpha, nelements(dg, cache))
+
+  Threads.@threads for element in eachelement(dg, cache)
+    indicator = indicator_threaded[Threads.threadid()]
+
+    # Calculate indicator variables at Gauss-Lobatto nodes
+    for j in eachnode(dg), i in eachnode(dg)
+      u_local = get_node_vars(u, equations, dg, i, j, element)
+      indicator[i, j] = löhner.variable(u_local, equations)
+    end
+
+    estimate = zero(real(dg))
+    for j in eachnode(dg), i in 2:nnodes(dg)-1
+      # x direction
+      u0 = indicator[i,   j]
+      up = indicator[i+1, j]
+      um = indicator[i-1, j]
+      estimate = max(estimate, löhner(um, u0, up))
+    end
+
+    for j in 2:nnodes(dg)-1, i in eachnode(dg)
+      # y direction
+      u0 = indicator[i, j, ]
+      up = indicator[i, j+1]
+      um = indicator[i, j-1]
+      estimate = max(estimate, löhner(um, u0, up))
+    end
+
+    # use the maximum as DG element indicator
+    alpha[element] = estimate
+  end
+
+  return alpha
+end
+
+# dirty Löhner estimate, direction by direction, assuming constant nodes
+@inline function (löhner::IndicatorLöhner)(um::Real, u0::Real, up::Real)
```

Comment:
I wouldn't call this OOP, honestly. How would you like to call this? `calc_local_löhner_estimate`? It was just convenient for me to write it as `löhner(um, u0, up)` when working with this indicator (and to avoid repeating the same nontrivial code multiple times).

Reply:
I've put the general question about OOP-ness on next week's Trixi meeting agenda.

Regarding this specific example: I am not sure I understand why using a regular method will require to repeat the code. The only downside I see is that you have to pass an additional argument (the `InidcatorLöhner` struct). However, by using a free function, it becomes possible to `grep` for its usage. I find it already challenging that many of the callbacks use callables instead of functions, but as I understand, it is required by the DiffEq.jl ecosphere, right?

However, I think for our general sanity it would be better if we ourselves use only regular functions, such that it is easy to find out how and where such a function is called.

Reply:
That's no problem at all and I'll be happy to use a simple function here. How would you like to call it? Is my suggestion above okay?
Re: duplicating code; I compared that to "not using a callable/function but the way it was before".

Reply:
Yep, the name sounds fine, although you could even drop the `calc_` and make it just `local_löhner_estimate`.

One thing I just realized... we are violating our ASCII-only policy with `Löhner`, aren't we? Shouldn't we call it `Loehner`?

Reply:
Fixed in bf63994.

We have `const IndicatorLoehner = IndicatorLöhner` and export that one, too, to make it easier for people with ASCII keyboard to use Trixi. As you know, I don't really like the strict ASCII policy that much :wink: Moreover, I don't like crippling the names of people that much and would prefer to keep `IndicatorLöhner`.

Reply:
Thanks. And thanks for thinking about me when creating IndicatorLöhner together with IndicatorLoehner :heart:!

Just to be consistent, shouldn't we also rename `flux_shima_etal` to `flux_島_etal`?

Reply:
Just kidding :-) Also, I don't even know which kanji is the correct one for that particular name...

Reply:
😅 I think it's okay to keep the current version - at least that's the name taken from the paper. But I'm open to add the "correct" one if you find a reference...